src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs

```
@@ -742,35 +743,163 @@ public override ClaimsPrincipal ValidateToken(string token, TokenValidationParam
+                    if (currentConfiguration != null)
+                    {
+                        validationParameters.ConfigurationManager.RequestRefresh();
+                        var lastConfig = currentConfiguration;
+                        currentConfiguration = validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).GetAwaiter().GetResult();
+
+                        // Only try to re-validate using the newly obtained config if it doesn't reference equal the previously used configuration.
+                        if (lastConfig != currentConfiguration)
+                            claimsPrincipal = outerToken != null ? ValidateJWE(token, outerToken, validationParameters, currentConfiguration, out signatureValidatedToken, out exceptionThrown) :
+                                ValidateJWS(token, validationParameters, currentConfiguration, out signatureValidatedToken, out exceptionThrown);
+                    }
+                }
+            }
+
+            if (claimsPrincipal != null)
+                return claimsPrincipal;
+
+            throw exceptionThrown;
+        }
+
+        private ClaimsPrincipal ValidateJWE(string decryptedJwt, JwtSecurityToken outerToken, TokenValidationParameters validationParameters, BaseConfiguration currentConfiguration, out SecurityToken signatureValidatedToken, out Exception exceptionThrown)
+        {
+            exceptionThrown = null;
+            signatureValidatedToken = null;
+            try
+            {
+                var innerToken = ValidateSignature(decryptedJwt, validationParameters, currentConfiguration);
+                outerToken.InnerToken = innerToken;
+                signatureValidatedToken = outerToken;
+                return ValidateTokenPayload(innerToken, validationParameters, currentConfiguration);
+            }
+            catch (Exception ex)
+            {
+                exceptionThrown = ex;
+                return null;
+            }
```

I'm not crazy about this sort of pattern in general (catching generic exceptions and then assuming the caller will always throw in the right way) but if we go down this route at least consider using ExceptionDispatchInfo instead. I believe the way the code is currently written, it will strip the stack trace info which I don't think is the intent and hurts debugging efforts.

[https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?redirectedfrom=MSDN&view=net-6.0#System_Runtime_ExceptionServices_ExceptionDispatchInfo_Capture_System_Exception](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?redirectedfrom=MSDN&view=net-6.0#System_Runtime_ExceptionServices_ExceptionDispatchInfo_Capture_System_Exception)_

---

I agree. We should not rethrow the exception.
We can throw a new exception, added the current exception as the inner.

---

Changed to exceptionDispatchInfo.Throw()

---

I like this a lot better than what was there before but it really begs the question for me, why are we even passing anything around for the exception, if we're just throwing in the method just above in the call stack why not just do a regular throw here instead of capturing it, letting control return to the caller and then immediately throwing?

---

Ah I see, it was too long since I looked at this last. It's to allow for the fallback in logic. This makes sense, this part LGTM.